roachtest: acceptance/bank/node-restart failed

SHA: https://github.com/cockroachdb/cockroach/commits/032c4980720abc1bdd71e4428e4111e6e6383297

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=acceptance/bank/node-restart PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1158877&tab=buildLog

```
The test failed on master:
	bank.go:394,bank.go:498,acceptance.go:78,test.go:1211: stall detected at round 5, no forward progress for 30s
	cluster.go:1152,bank.go:248: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod start teamcity-1158877-acceptance:4 returned:
		stderr:
		
		stdout:
		teamcity-1158877-acceptance: starting.............................: signal: killed

```

Running on https://github.com/cockroachdb/cockroach/commit/bd27eb358f558bb7598945318240335ebcfcdf13 I ran into the following:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5e70446]

goroutine 68 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0008367e0, 0x74e0620, 0xc000956270)
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:181 +0x121
panic(0x6983380, 0x92382e0)
        /usr/local/Cellar/go/1.12.7/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/util/quotapool.(*IntAlloc).Freeze(0x0)
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/util/quotapool/intpool.go:72 +0x26
main.(*cluster).Save(0xc000518780, 0x74e0620, 0xc0009562a0, 0xc0002a45a0, 0x4f, 0xc0000d0960)
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:1067 +0xb7
main.(*testRunner).runWorker(0xc00022af00, 0x74e0620, 0xc000956270, 0xc0002062fc, 0x2, 0xc0009561b0, 0xc00027aa40, 0xc0004cc420, 0xc000956201, 0x6d1829e, ...)
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:478 +0xb26
main.(*testRunner).Run.func2(0x74e0620, 0xc000956270)
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:274 +0x226
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00027aa70, 0xc0008367e0, 0xc00022e0e0)
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0xfb
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
        /Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xa8
```

This then prompts the `dead node detection` failure as seen by the roachtest.

--- a/pkg/cmd/roachtest/bank.go
+++ b/pkg/cmd/roachtest/bank.go
@@ -57,7 +57,7 @@ func (client *bankClient) transferMoney(ctx context.Context, numAccounts, maxTra
 	// statement timeout, which unfortunately precludes the use of prepared
 	// statements.
 	q := fmt.Sprintf(`
-SET statement_timeout = '31s';
+SET statement_timeout = '30s';
 UPDATE bank.accounts
    SET balance = CASE id WHEN %[1]d THEN balance-%[3]d WHEN %[2]d THEN balance+%[3]d END
  WHERE id IN (%[1]d, %[2]d) AND (SELECT balance >= %[3]d FROM bank.accounts WHERE id = %[1]d);
@@ -237,9 +237,7 @@ func (s *bankState) startChaosMonkey(
 					break
 				}
 				t.l.Printf("round %d: restarting %d\n", curRound, i)
-
-				c.Stop(ctx, c.Node(i))
-				c.Start(ctx, t, c.Node(i))
+				c.Restart(ctx, t, c.Node(i))
 			}
 
 			preCount := s.counts()
@@ -438,7 +436,7 @@ func runBankClusterRecovery(ctx context.Context, t *test, c *cluster) {
 	}
 	s.startChaosMonkey(ctx, t, c, pickNodes, -1)
 
-	s.waitClientsStop(ctx, t, c, 30*time.Second)
+	s.waitClientsStop(ctx, t, c, 45*time.Second)
 
 	// Verify accounts.
 	s.verifyAccounts(ctx, t)
@@ -480,7 +478,7 @@ func runBankNodeRestart(ctx context.Context, t *test, c *cluster) {
 	}
 	s.startChaosMonkey(ctx, t, c, pickNodes, clientIdx)
 
-	s.waitClientsStop(ctx, t, c, 30*time.Second)
+	s.waitClientsStop(ctx, t, c, 45*time.Second)
 
 	// Verify accounts.
 	s.verifyAccounts(ctx, t)
@@ -511,7 +509,7 @@ func runBankNodeZeroSum(ctx context.Context, t *test, c *cluster) {
 	}
 
 	s.startSplitMonkey(ctx, 2*time.Second, c)
-	s.waitClientsStop(ctx, t, c, 30*time.Second)
+	s.waitClientsStop(ctx, t, c, 45*time.Second)
 
 	s.verifyAccounts(ctx, t)
 
@@ -559,7 +557,7 @@ func runBankZeroSumRestart(ctx context.Context, t *test, c *cluster) {
 	// Starting up the goroutines that restart and do splits and lease moves.
 	s.startChaosMonkey(ctx, t, c, pickNodes, -1)
 	s.startSplitMonkey(ctx, 2*time.Second, c)
-	s.waitClientsStop(ctx, t, c, 30*time.Second)
+	s.waitClientsStop(ctx, t, c, 45*time.Second)
 
 	// Verify accounts.
 	s.verifyAccounts(ctx, t)

--- a/pkg/cmd/roachtest/cluster.go
+++ b/pkg/cmd/roachtest/cluster.go
@@ -375,6 +375,12 @@ func execCmd(ctx context.Context, l *logger, args ...string) error {
 	if err := cmd.Run(); err != nil {
 		cancel()
 		wg.Wait() // synchronize access to ring buffer
+
+		// Context deadline exceeded errors opaquely appear as "signal killed" when
+		// manifested. We surface this error explicitly.
+		if ctx.Err() == context.DeadlineExceeded {
+			return ctx.Err()
+		}
 		return errors.Wrapf(
 			err,
 			"%s returned:\nstderr:\n%s\nstdout:\n%s",
@@ -1633,6 +1639,26 @@ func roachprodArgs(opts []option) []string {
 		args = append(args, ([]string)(a)...)
 	}
 	return args
+}
+
+// Restart restarts the specified cockroach node. It takes a test and, on error,
+// calls t.Fatal().
+func (c *cluster) Restart(ctx context.Context, t *test, node nodeListOption) {
+	// We bound the time taken to restart a node through roachprod. Because
+	// roachprod uses SSH, it's particularly vulnerable to network flakiness (as
+	// seen in #35326) and may stall indefinitely. Setting up timeouts better
+	// surfaces this kind of failure.
+	//
+	// TODO(irfansharif): The underlying issue here is the fact that we're running
+	// roachprod commands that may (reasonably) fail due to connection issues, and
+	// we're unable to retry them safely (the underlying commands are
+	// non-idempotent). Presently we simply fail the entire test, when really we
+	// should be able to retry the specific roachprod commands.
+	var cancel func()
+	ctx, cancel = context.WithTimeout(ctx, 30*time.Second)
+	c.Stop(ctx, node)
+	c.Start(ctx, t, node)
+	cancel()
 }
 
 // StartE starts cockroach nodes on a subset of the cluster. The nodes parameter

